java-checks/src/main/java/org/sonar/java/checks/regex/GraphemeClustersInClassesCheck.java

```diff
@@ -0,0 +1,85 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.regex;
+
+import java.text.Normalizer;
+import java.util.Collections;
+import javax.annotation.Nullable;
+import org.sonar.check.Rule;
+import org.sonar.java.regex.RegexParseResult;
+import org.sonar.java.regex.ast.CharacterClassTree;
+import org.sonar.java.regex.ast.CharacterClassUnionTree;
+import org.sonar.java.regex.ast.JavaCharacter;
+import org.sonar.java.regex.ast.PlainCharacterTree;
+import org.sonar.java.regex.ast.RegexBaseVisitor;
+import org.sonar.java.regex.ast.RegexTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+
+@Rule(key = "S5868")
+public class GraphemeClustersInClassesCheck extends AbstractRegexCheck {
+
+  private static final String MESSAGE = "Extract these Grapheme Clusters from the character class.";
+
+  @Override
+  public void checkRegex(RegexParseResult regexForLiterals, MethodInvocationTree mit) {
+    new GraphemeInClassVisitor().visit(regexForLiterals);
+  }
+
+  private class GraphemeInClassVisitor extends RegexBaseVisitor {
+
+    boolean containGrapheme = false;
+
+    @Override
+    public void visitCharacterClass(CharacterClassTree tree) {
+      super.visitCharacterClass(tree);
+      if (containGrapheme) {
+        reportIssue(tree, MESSAGE, null, Collections.emptyList());
+      }
+      containGrapheme = false;
+    }
+
+    @Override
+    public void visitCharacterClassUnion(CharacterClassUnionTree tree) {
+      JavaCharacter previousChar = null;
+      for (RegexTree child : tree.getCharacterClasses()) {
+        if (child.is(RegexTree.Kind.PLAIN_CHARACTER)) {
+          JavaCharacter currentCharacter = ((PlainCharacterTree) child).getContents();
+          if (!currentCharacter.isEscapedUnicode()) {
+            char currentChar = currentCharacter.getCharacter();
+            if (canCanonicallyCombine(previousChar, currentChar)) {
+              containGrapheme = true;
+            }
+            previousChar = currentCharacter;
+            continue;
+          }
+        }
+        previousChar = null;
+      }
+
+      super.visitCharacterClassUnion(tree);
+    }
+
+    private boolean canCanonicallyCombine(@Nullable JavaCharacter previousChar, char currentChar) {
+      return previousChar != null
+        && Normalizer.normalize(new StringBuilder().append(previousChar.getCharacter()).append(currentChar), Normalizer.Form.NFC).length() == 1;
```

Comment:
I feel it is not enough.

I was surprised to see that a simpler version of the test case from the `Noncompliant` part of the RSPEC was marked as Compliant, while it shouldn't. The following code, when executed:

```
"cc̈c".replaceAll("[c̈]", "X");
```
would print XXXX, instead of cXc, and it should be reported!